formats/json/commonMain/src/kotlinx/serialization/json/internal/StreamingJsonDecoder.kt

```diff
@@ -33,7 +34,37 @@ internal open class StreamingJsonDecoder(
     override fun decodeJsonElement(): JsonElement = JsonTreeReader(json.configuration, lexer).read()
 
     override fun <T> decodeSerializableValue(deserializer: DeserializationStrategy<T>): T {
-        return decodeSerializableValuePolymorphic(deserializer)
+        /*
+         * This is an optimized path over decodeSerializableValuePolymorphic(deserializer):
+         * dSVP reads the very next JSON tree into a memory as JsonElement and then runs TreeJsonDecoder over it
+         * in order to deal with an arbitrary order of keys, but with the price of additional memory pressure
+         * and CPU consumption.
+         * We would like to provide best possible performance for data produced by kotlinx.serialization
+         * itself, for that we do the following optimistic optimization:
+         *
+         * 0) Remember current position in the string
+         * 1) Read the very next key of JSON structure
+         * 2) If it matches*  the descriminator key, read the value, remember current position
+         * 3) Return the value, recover an initial position
+         * 4) Right after starting the structure beginning, immediately skip over discriminator
+         *    using position from '2'.
+         * In such scenario we neither process the same input twice nor create aux data structures.
+         *
+         * (*) -- if it doesn't match, fallback to dSVP method.
+         */
+        if (deserializer !is AbstractPolymorphicSerializer<*> || json.configuration.useArrayPolymorphism) {
```

Comment:
@sandwwraith PTAL.

This is a working prototype with actual measurements, please take a look and see if the overall approach is safe and sound. It still requires polishing (e.g. I'm pretty sure the current version is incorrect for IO steaming, we just don't have enough test coverage yet) which I'm eager to do as soon as we'll agree on the overall approach